index.js

```diff
@@ -54,11 +54,13 @@ function fastifyWebsocket (fastify, opts, next) {
     }
   })
 
-  const handleUpgrade = (rawRequest, callback) => {
+  const handleUpgrade = (rawRequest, request, callback) => {
     wss.handleUpgrade(rawRequest, rawRequest[kWs], rawRequest[kWsHead], (socket) => {
       wss.emit('connection', socket, rawRequest)
-
       const connection = WebSocket.createWebSocketStream(socket, opts.connectionOptions)
+      socket.afterDuplex = true
+      socket.validator = request && request.context.schema ? fastify.validatorCompiler({ schema: request.context.schema.body }) : null
```

Comment:
It should move to `onRoute` hook. It compiles on every request which is not optimal.
```js
// onRoute hook
const kWebSocketSchema = Symbol('websocket-body-schema')
requset.context[kWebSocketSchema] = request && request.context.schema ? fastify.validatorCompiler({ schema: request.context.schema.body }) : null

// here
socket.validator = requset.context[kWebSocketSchema]
```